src/components/include/cbuf.h

```diff
@@ -343,121 +350,198 @@ PERCPU_EXTERN(cbuf_alloc_freelists);
 static inline struct cbuf_meta *
 __cbuf_freelist_get(int size)
 {
+	/* We have per-core, per-component, per-size free-list */
 	struct cbuf_freelist *fl = PERCPU_GET(cbuf_alloc_freelists);
 	int order = ones(size-1) - PAGE_ORDER;
 	struct cbuf_meta *m;
+	if (!(pow2(size) && size >= PAGE_SIZE)) {
+		printc("cbuf free size %d spd %d\n", size, cos_spd_id());
+	}
 	assert(pow2(size) && size >= PAGE_SIZE);
 	assert(order >= 0 && order < WORD_SIZE-PAGE_ORDER);
 	m = &(fl->freelist_head[order]);
-	if (unlikely(!m->next)) m->next = m;
+	assert(m->next);
 	return m;
 }
 
-static inline void *
-cbuf_alloc_ext(unsigned int sz, cbuf_t *cb, unsigned int flag)
+/* Precondition: "next" pointer is not NULL. 
+ * We can reuse a cbuf from two source.
+ * 1. from local freelist
+ * 2. from garbage collection
+ * The manager guarantees the precondition for case 2.
+ *
+ * Before next pointer is set to NULL, manager will not 
+ * collect it. See cbuf_collect for details.
+ * As other clients or threads have no reference to this 
+ * cbuf, they will not access the meta concurrently.
+ * But the manager may modify the meta at the same time.
+ * So those atomic instructions are necessary*/
+static inline int 
+__cbuf_try_take(struct cbuf_meta *cm, unsigned int flag)
 {
-	void *ret;
-	int cbid, len, already_used, inconsistent, mapped_in, flags;
-	struct cbuf_meta *cm, *fl, old_head, new_head;
-	unsigned long long *target, *old, *update;
-
-	if (unlikely(flag & CBUF_EXACTSZ)) {
-		cm   = __cbuf_alloc_slow(sz, &len, flag);
-		assert(cm);
-		goto done;
+	int inconsistent, r = 0;
+	unsigned long old_nfo, new_nfo;
+	assert(cm && cm->next);
+	old_nfo = cm->nfo;
+	assert(!(old_nfo & CBUF_REFCNT_MAX));
+
+	inconsistent = old_nfo & CBUF_INCONSISENT;
+	if (unlikely(inconsistent)) {
+		/* It has been or is going to be taken away by the 
+		 * manager. We will not leak cbuf here.
+		 * Do not modify other fields! */
+		CBUF_FLAG_REM(cm, CBUF_INCONSISENT);
+		goto ret;
 	}
 
-	sz = nlepow2(round_up_to_page(sz));
+	new_nfo = old_nfo+1;
+	if (unlikely(flag & CBUF_TMEM)) new_nfo |= CBUF_TMEM;
+	else new_nfo &= ~CBUF_TMEM;
+	if (unlikely(!cos_cas(&cm->nfo, old_nfo, new_nfo))) {
+		/* This failure maybe because:
+		 * 1. manager tries to take this cbuf away, set inconsistent
+		 * 2. manager set/unset relinq bit.
+		 * For case 2, this cbuf can be collected later.
+		 * No cbuf leak */
+		if (CBUF_INCONSISENT(cm)) CBUF_FLAG_REM(cm, CBUF_INCONSISENT);
+		goto ret;
+	}
+	r = 1;
+	/* guarantee the order between CAS and set "next".
+	 * If the next is set to NULL before increment refcnt, the
+	 * manager may think this cbuf is garbage and collect it. */
+	/* But there is dependency between those two, we do not 
+	 * need explicit barrier. */
```

Comment:
I remember there is a control flow dependency due to the if statement.
Maybe it is different for different compilers